flowmachine/flowmachine/core/server/action_handlers.py

```diff
@@ -0,0 +1,327 @@
+
+    Returns a dict with all supported flowmachine queries as keys
+    and the associated schema for the query parameters as values.
+    """
+    spec = APISpec(
+        title="FlowAPI",
+        version="1.0.0",
+        openapi_version="3.0.2",
+        plugins=[MarshmallowPlugin()],
+    )
+    spec.components.schema("FlowmachineQuerySchema", schema=FlowmachineQuerySchema)
+    schemas_spec = spec.to_dict()["components"]["schemas"]
+    schemas_spec = sort_recursively(schemas_spec)  # needed for reproducible output
+    return ZMQReply(status="success", payload={"query_schemas": schemas_spec})
+
+
+def action_handler__run_query(**action_params):
+    """
+    Handler for the 'run_query' action.
+
+    Constructs a flowmachine query object, sets it running and returns the query_id.
+    For this action handler the `action_params` are exactly the query kind plus the
+    parameters needed to construct the query.
+    """
+    try:
+        query_obj = FlowmachineQuerySchema().load(action_params)
+    except ValidationError as exc:
+        # The dictionary of marshmallow errors can contain integers as keys,
+        # which will raise an error when converting to JSON (where the keys
+        # must be strings). Therefore we transform the keys to strings here.
+        error_msg = "Parameter validation failed."
+        validation_error_messages = convert_dict_keys_to_strings(exc.messages)
+        return ZMQReply(
+            status="error", msg=error_msg, payload=validation_error_messages
+        )
+
+    q_info_lookup = QueryInfoLookup(Query.redis)
+    try:
+        query_id = q_info_lookup.get_query_id(action_params)
+    except QueryInfoLookupError:
+        # Set the query running (it's safe to call this even if the query was set running before)
+        query_id = query_obj.store_async()
+
+        # Register the query as "known" (so that we can later look up the query kind
+        # and its parameters from the query_id).
+
+        q_info_lookup.register_query(query_id, action_params)
+
+    return ZMQReply(status="success", payload={"query_id": query_id})
+
+
+def _get_query_kind_for_query_id(query_id):
+    """
+    Helper function to look up the query kind corresponding to the
+    given query id. Returns `None` if the query_id does not exist.
+
+    Parameters
+    ----------
+    query_id : str
+        Identifier of the query.
+
+    Returns
+    -------
+    str or None
+        The query kind associated with this query_id (or None
+        if no query with this query_id exists).
+    """
+    q_info_lookup = QueryInfoLookup(Query.redis)
+    try:
+        return q_info_lookup.get_query_kind(query_id)
+    except UnkownQueryIdError:
+        return None
+
+
+def action_handler__poll_query(query_id):
+    """
+    Handler for the 'poll_query' action.
+
+    Returns the status of the query with the given `query_id`.
+    """
+    query_kind = _get_query_kind_for_query_id(query_id)
```

Comment:
One thing to consider for the future is whether this is actually something that should go via zmq as well, particularly if we make all this a bit more more async.

Reply:
Not sure I see what exactly you have in mind. Can you elaborate?